plugin/namespaces.go

```diff
@@ -0,0 +1,84 @@
+package plugin
+
+import (
+	"github.com/mitchellh/cli"
+	"strings"
+)
+
+// TODO: Everything in this file needs better naming.
+type namespaces struct {
+	displays []string
+	plugins  []string
+}
+
+func newNamespaces(pluginCommands map[string]string, coreCommands map[string]cli.CommandFactory, hiddenCommands []string) namespaces {
+	cores := keysForCommandFactoryMap(coreCommands)
+	coreDisplays := missing(hiddenCommands, cores)
+	plugins := keysForStringMap(pluginCommands)
+
+	return namespaces{
+		displays: append(coreDisplays, plugins...),
+		plugins:  plugins,
+	}
+}
+
+func (n *namespaces) forPlugins() []string {
+	var namespaces []string
+	for _, name := range n.plugins {
+		segments := strings.Split(name, " ")
+
+		for i, _ := range make([]int, len(segments)) {
+			namespace := strings.Join(segments[:i+1], " ")
+			namespaces = append(namespaces, namespace)
+		}
+	}
+
+	return missing(n.displays, namespaces)
+}
+
+// TODO: Question: Any way to accept both `map[string]string` and `map[string]cli.CommandFactory` at the same time?
+func keysForStringMap(m map[string]string) []string {
+	keys := make([]string, len(m))
+
+	i := 0
+	for k := range m {
```

Comment:
```go
for i, k:= range m {
  keys[i] = k
}
```

Should handle the index without doing it manually

Reply:
Because we want to turn the map into a slice, and slice keys must be intergers.

In `for i, k:= range m {`, `i` is the string key we want to collect, `k` is the vaule we dont need.

See: https://play.golang.org/p/_-rKpEraZt6

Reply:
Ah sorry, I might do this then:

```go
keys := []string{}

for k := range m {
  keys = append(keys, k)
}

return keys
```

It reads better to me and is a little simpler?

Reply:
This snippet no longer exist after refactoring.